File: codecamp/src/main/java/com/gdogaru/codecamp/view/agenda/calendar/CalendarFragment.java
package com.gdogaru.codecamp.view.agenda.calendar;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.gdogaru.codecamp.di.Injectable;
import com.gdogaru.codecamp.model.Schedule;
import com.gdogaru.codecamp.model.Session;
import com.gdogaru.codecamp.model.Track;
import com.gdogaru.codecamp.svc.CodecampClient;
import com.gdogaru.codecamp.util.Joiner;
import com.gdogaru.codecamp.view.agenda.SessionsFragment;
import com.gdogaru.codecamp.view.session.SessionExpandedActivity;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import icepick.State;

public class CalendarFragment extends SessionsFragment implements Injectable {

    private static final Comparator<? super Session> SESSION_BY_DATE_COMPARATOR = (Comparator<Session>) (lhs, rhs) -> lhs.getStartTime().compareTo(rhs.getStartTime());

    ArrayList<Integer> sessIds = new ArrayList<Integer>();
    Timer currentTimer;
    @Inject
    CodecampClient codecampClient;
    @State
    Calendar.CalendarState calendarState;
    private int offset;
    private Calendar calendar;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        calendar = new Calendar(getActivity());
        return calendar;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        updateDisplay();
    }

    @Override
    public void onResume() {
        super.onResume();
        currentTimer = new Timer();
        currentTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                calendar.post(() -> calendar.updateCurrentTime(DateTime.now()));
            }
        }, 500, 30_000);

        calendar.setBookmarked(bookmarkingService.getBookmarked(codecampClient.getEvent().getTitle()));
        if (calendarState != null) {
            calendar.postDelayed(() -> calendar.setState(calendarState), 300);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (currentTimer != null) {
            currentTimer.cancel();
        }
        calendarState = calendar.getState();
    }

    public void updateDisplay() {
        Schedule schedule = codecampClient.getSchedule();
        List<Session> sessions = new ArrayList<>(schedule.getSessions());
        List<Track> tracks = new ArrayList<>(schedule.getTracks());
        Set<String> bookmarked = bookmarkingService.getBookmarked(codecampClient.getEvent().getTitle());

        if (getFavoritesOnly()) {
            keepFavoritesOnly(sessions, tracks, codecampClient.getEvent().getTitle(), bookmarked);
        }

        List<CEvent> events = new ArrayList<>();
        Collections.sort(sessions, SESSION_BY_DATE_COMPARATOR);
        for (int i = 0; i < sessions.size(); i++) {
            Session ss = sessions.get(i);
            int preferedIdx = 0;
            if (ss.getTrack() != null) {
                Track track = getTrack(tracks, ss.getTrack());
                if (track != null) preferedIdx = track.getDisplayOrder();
            }
            String descLine2 = ss.getTrack();
            events.add(new CEvent(ss.getId(), ss.getStartTime(), ss.getEndTime(), preferedIdx, ss.getTitle(),
                    createSpeakerName(ss),
                    descLine2 == null ? "" : descLine2));
        }
        initSessionIds(sessions);
        calendar.setCurrentTime(DateTime.now());
        calendar.setEvents(events);
        calendar.setScheduleDate(schedule.getDate());

        calendar.setEventListener(event -> displayEventDetails(event.event.id));
    }

    private void keepFavoritesOnly(List<Session> sessions, List<Track> tracks, String eventId, Set<String> bookmarked) {
        for (Iterator<Session> iterator = sessions.iterator(); iterator.hasNext(); ) {
            Session s = iterator.next();
            if (!bookmarked.contains(s.getId())) {
                iterator.remove();
            }
        }
        Set<String> rt = new HashSet<>();
        for (Session s : sessions) rt.add(s.getTrack());
        for (Iterator<Track> iterator = tracks.iterator(); iterator.hasNext(); ) {
            Track t = iterator.next();
            if (!rt.contains(t.getName())) iterator.remove();
        }
    }

    private Track getTrack(List<Track> tracks, String track) {
        for (Track t : tracks) {
            if (t.getName().equals(track)) {
                return t;
            }
        }
        return null;
    }


    private void initSessionIds(List<Session> sessions) {
        List<Session> ss = new ArrayList<>(sessions);
        sessIds = new ArrayList<>();
        for (int i = 0; i < ss.size(); i++) {
            sessIds.add(i);
        }
    }

    private String createSpeakerName(Session session) {
        if (session.getSpeakerIds() == null || session.getSpeakerIds().isEmpty()) return "";
        return Joiner.on(", ").join(session.getSpeakerIds());
    }

    private void displayEventDetails(String id) {
        SessionExpandedActivity.start(getActivity(), id, null);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        //first saving my state, so the bundle wont be empty.
        //http://code.google.com/p/android/issues/detail?id=19917
        outState.putLong("something", 1);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }
}

File: codecamp/src/main/java/com/gdogaru/codecamp/view/agenda/calendar/CalendarFragment.kt
/*
 * Copyright (c) 2019 Gabriel Dogaru - gdogaru@gmail.com
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.gdogaru.codecamp.view.agenda.calendar


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import com.evernote.android.state.State
import com.gdogaru.codecamp.api.model.Schedule
import com.gdogaru.codecamp.api.model.Session
import com.gdogaru.codecamp.di.Injectable
import com.gdogaru.codecamp.view.agenda.AgendaFragmentDirections
import com.gdogaru.codecamp.view.agenda.AbstractSessionsListFragment
import com.gdogaru.codecamp.view.agenda.calendar.component.CEvent
import com.gdogaru.codecamp.view.agenda.calendar.component.Calendar
import com.gdogaru.codecamp.view.agenda.calendar.component.DisplayEvent
import com.gdogaru.codecamp.view.agenda.calendar.component.EventListener
import org.threeten.bp.LocalDateTime
import java.util.*
import javax.inject.Inject


class CalendarFragment : AbstractSessionsListFragment(), Injectable {

    private var sessIds = ArrayList<Int>()
    private var currentTimer: Timer? = null
    private val offset: Int = 0
    private lateinit var calendar: Calendar
    @State
    lateinit var calendarState: Calendar.CalendarState
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: CalendarFragmentViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(requireActivity(), viewModelFactory).get(CalendarFragmentViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        calendar = Calendar(requireActivity())
        return calendar
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.eventList().observe(this, androidx.lifecycle.Observer { s ->
            s?.let { updateDisplay(it.first, it.second) }
        })
    }


    override fun onResume() {
        super.onResume()
        currentTimer = Timer()
        currentTimer!!.schedule(object : TimerTask() {
            override fun run() {
                calendar.post { calendar.updateCurrentTime(LocalDateTime.now()) }
            }
        }, 500, 30000)
    }

    override fun onPause() {
        super.onPause()
        if (currentTimer != null) {
            currentTimer!!.cancel()
        }
        calendarState = calendar.state
    }

    override fun setFavoritesOnly(favoritesOnly: Boolean) {
        viewModel.setFavoritesOnly(favoritesOnly)
    }

    private fun updateDisplay(events: List<CEvent>, schedule: Schedule) {
        initSessionIds(schedule.sessions)
        calendar.setCurrentTime(LocalDateTime.now())
        calendar.setEvents(events)
        calendar.setScheduleDate(schedule.date)

        calendar.setEventListener(object : EventListener {
            override fun eventCLicked(event: DisplayEvent) {
                displayEventDetails(event.event.id)
            }
        })
    }

    private fun initSessionIds(sessions: List<Session>) {
        val ss = ArrayList(sessions)
        sessIds = ArrayList()
        for (i in ss.indices) {
            sessIds.add(i)
        }
    }


    private fun displayEventDetails(id: String) {
        findNavController().navigate(AgendaFragmentDirections.showSessionInfo(id))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        //first saving my state, so the bundle wont be empty.
        //http://code.google.com/p/android/issues/detail?id=19917
        outState.putLong("something", 1)
        super.onSaveInstanceState(outState)
    }

}
